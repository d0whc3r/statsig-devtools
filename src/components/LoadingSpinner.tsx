interface LoadingSpinnerProps {
  size?: 'sm' | 'md' | 'lg'
  message?: string
  className?: string
}

export function LoadingSpinner({ size = 'md', message, className = '' }: LoadingSpinnerProps) {
  const sizeClasses = {
    sm: 'w-4 h-4',
    md: 'w-8 h-8',
    lg: 'w-12 h-12',
  }

  return (
    <div className={`flex flex-col items-center justify-center ${className}`} data-testid="loading-spinner">
      <div
        className={`animate-spin rounded-full border-2 border-slate-200 ${sizeClasses[size]}`}
        style={{
          borderTopColor: '#1e40af',
          borderRightColor: '#3b82f6',
          background: 'conic-gradient(from 0deg, transparent, rgba(30, 64, 175, 0.1))',
        }}
      />
      {message && <p className="mt-3 text-sm font-medium text-slate-600">{message}</p>}
    </div>
  )
}

export default LoadingSpinner
