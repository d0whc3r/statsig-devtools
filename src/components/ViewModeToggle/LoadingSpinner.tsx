import React from 'react'

import type { ViewMode } from '../../services/view-mode'

interface LoadingSpinnerProps {
  currentMode: ViewMode
}

/**
 * Loading spinner component for view mode toggle
 */
export function LoadingSpinner({ currentMode }: LoadingSpinnerProps) {
  const spinnerSize = currentMode === 'popup' ? 'h-3 w-3' : 'h-4 w-4'

  return (
    <svg className={`animate-spin ${spinnerSize}`} fill="none" viewBox="0 0 24 24">
      <circle className="opacity-25" cx="12" cy="12" r="10" stroke="currentColor" strokeWidth="4" />
      <path
        className="opacity-75"
        fill="currentColor"
        d="M4 12a8 8 0 018-8V0C5.373 0 0 5.373 0 12h4zm2 5.291A7.962 7.962 0 014 12H0c0 3.042 1.135 5.824 3 7.938l3-2.647z"
      />
    </svg>
  )
}
