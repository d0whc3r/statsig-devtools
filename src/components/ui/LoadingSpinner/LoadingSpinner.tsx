import { useViewMode } from '@/src/hooks/useViewMode'
import { cn } from '@/src/utils/cn'

export function LoadingSpinner() {
  const { viewMode } = useViewMode()
  return (
    <div
      role="status"
      aria-label="Loading..."
      className={cn(
        'animate-spin rounded-full border-4 border-gray-200 border-t-blue-600',
        viewMode === 'popup' && 'h-8 w-8',
        viewMode === 'sidebar' && 'h-12 w-12',
        viewMode === 'tab' && 'h-16 w-16',
      )}
    />
  )
}
