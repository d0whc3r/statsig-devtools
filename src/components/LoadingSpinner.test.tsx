import { describe, expect, it } from 'vitest'

import { LoadingSpinner } from './LoadingSpinner'

import { render, screen, within } from '@testing-library/react'

describe('LoadingSpinner', () => {
  it('should match snapshot with default props', () => {
    const { container } = render(<LoadingSpinner />)
    expect(container).toMatchSnapshot()
  })

  describe('basic rendering', () => {
    it('should render spinner with default props', () => {
      render(<LoadingSpinner />)

      const container = screen.getByTestId('loading-spinner')
      expect(container).toBeInTheDocument()
      expect(container).toHaveClass('flex', 'flex-col', 'items-center', 'justify-center')

      // Check for spinner element using within
      const spinner = within(container).getByRole('generic')
      expect(spinner).toBeInTheDocument()
      expect(spinner).toHaveClass('animate-spin', 'rounded-full', 'border-2', 'border-slate-200')
    })

    it('should not render message when not provided', () => {
      render(<LoadingSpinner />)

      expect(screen.getByTestId('loading-spinner')).toBeInTheDocument()
      expect(screen.queryByText(/loading/i)).not.toBeInTheDocument()
    })
  })

  describe('size variants', () => {
    it('should render small spinner correctly', () => {
      render(<LoadingSpinner size="sm" />)

      const container = screen.getByTestId('loading-spinner')
      const spinner = within(container).getByRole('generic')
      expect(spinner).toHaveClass('w-4', 'h-4')
    })

    it('should render medium spinner correctly (default)', () => {
      render(<LoadingSpinner size="md" />)

      const container = screen.getByTestId('loading-spinner')
      const spinner = within(container).getByRole('generic')
      expect(spinner).toHaveClass('w-8', 'h-8')
    })

    it('should render large spinner correctly', () => {
      render(<LoadingSpinner size="lg" />)

      const container = screen.getByTestId('loading-spinner')
      const spinner = within(container).getByRole('generic')
      expect(spinner).toHaveClass('w-12', 'h-12')
    })

    it('should default to medium size when size prop is not provided', () => {
      render(<LoadingSpinner />)

      const container = screen.getByTestId('loading-spinner')
      const spinner = within(container).getByRole('generic')
      expect(spinner).toHaveClass('w-8', 'h-8')
    })
  })

  describe('message display', () => {
    it('should render message when provided', () => {
      const testMessage = 'Loading data...'
      render(<LoadingSpinner message={testMessage} />)

      const message = screen.getByText(testMessage)
      expect(message).toBeInTheDocument()
      expect(message).toHaveClass('mt-3', 'text-sm', 'font-medium', 'text-slate-600')
    })

    it('should render long message correctly', () => {
      const longMessage = 'This is a very long loading message that should still be displayed correctly'
      render(<LoadingSpinner message={longMessage} />)

      const message = screen.getByText(longMessage)
      expect(message).toBeInTheDocument()
      expect(message).toHaveClass('mt-3', 'text-sm', 'font-medium', 'text-slate-600')
    })

    it('should render empty string message', () => {
      render(<LoadingSpinner message="" />)

      // Empty string should not render a paragraph element
      expect(screen.queryByRole('paragraph')).not.toBeInTheDocument()
    })

    it('should handle special characters in message', () => {
      const specialMessage = 'Loading... 50% (10/20) âœ“'
      render(<LoadingSpinner message={specialMessage} />)

      const message = screen.getByText(specialMessage)
      expect(message).toBeInTheDocument()
    })
  })

  describe('custom className', () => {
    it('should apply custom className to container', () => {
      const customClass = 'my-custom-class'
      render(<LoadingSpinner className={customClass} />)

      const container = screen.getByTestId('loading-spinner')
      expect(container).toHaveClass(customClass)
      // Should still have default classes
      expect(container).toHaveClass('flex', 'flex-col', 'items-center', 'justify-center')
    })

    it('should apply multiple custom classes', () => {
      const customClasses = 'custom-class-1 custom-class-2 p-4'
      render(<LoadingSpinner className={customClasses} />)

      const container = screen.getByTestId('loading-spinner')
      expect(container).toHaveClass('custom-class-1', 'custom-class-2', 'p-4')
    })

    it('should handle empty className gracefully', () => {
      render(<LoadingSpinner className="" />)

      const container = screen.getByTestId('loading-spinner')
      expect(container).toHaveClass('flex', 'flex-col', 'items-center', 'justify-center')
    })

    it('should default to empty string when className is not provided', () => {
      render(<LoadingSpinner />)

      const container = screen.getByTestId('loading-spinner')
      expect(container).toHaveClass('flex', 'flex-col', 'items-center', 'justify-center')
    })
  })

  describe('combined props', () => {
    it('should render correctly with all props provided', () => {
      const props = {
        size: 'lg' as const,
        message: 'Loading large content...',
        className: 'my-spinner-class',
      }

      render(<LoadingSpinner {...props} />)

      const container = screen.getByTestId('loading-spinner')
      expect(container).toHaveClass('my-spinner-class', 'flex', 'flex-col', 'items-center', 'justify-center')

      const spinner = within(container).getByRole('generic')
      expect(spinner).toHaveClass('w-12', 'h-12')

      const message = screen.getByText('Loading large content...')
      expect(message).toBeInTheDocument()
    })

    it('should render small spinner with message and custom class', () => {
      render(<LoadingSpinner size="sm" message="Small loading..." className="compact" />)

      const container = screen.getByTestId('loading-spinner')
      expect(container).toHaveClass('compact')

      const spinner = within(container).getByRole('generic')
      expect(spinner).toHaveClass('w-4', 'h-4')

      const message = screen.getByText('Small loading...')
      expect(message).toBeInTheDocument()
    })
  })

  describe('accessibility', () => {
    it('should be accessible to screen readers', () => {
      render(<LoadingSpinner message="Loading content" />)

      const container = screen.getByTestId('loading-spinner')
      expect(container).toBeInTheDocument()

      // Message should be readable by screen readers
      const message = screen.getByText('Loading content')
      expect(message).toBeInTheDocument()
    })

    it('should work without message for screen readers', () => {
      render(<LoadingSpinner />)

      const container = screen.getByTestId('loading-spinner')
      expect(container).toBeInTheDocument()
      // Should not have any text content when no message is provided
      expect(container).not.toHaveTextContent(/./)
    })
  })

  describe('spinner animation classes', () => {
    it('should have correct animation and styling classes', () => {
      render(<LoadingSpinner />)

      const container = screen.getByTestId('loading-spinner')
      const spinner = within(container).getByRole('generic')
      expect(spinner).toHaveClass('animate-spin', 'rounded-full', 'border-2', 'border-slate-200')
    })

    it('should maintain animation classes across different sizes', () => {
      const { rerender } = render(<LoadingSpinner size="sm" />)

      let container = screen.getByTestId('loading-spinner')
      let spinner = within(container).getByRole('generic')
      expect(spinner).toHaveClass('animate-spin', 'border-2', 'border-slate-200')

      rerender(<LoadingSpinner size="lg" />)

      container = screen.getByTestId('loading-spinner')
      spinner = within(container).getByRole('generic')
      expect(spinner).toHaveClass('animate-spin', 'border-2', 'border-slate-200')
    })
  })

  describe('edge cases', () => {
    it('should handle undefined props gracefully', () => {
      const props = {
        size: undefined as any,
        message: undefined,
        className: undefined as any,
      }

      render(<LoadingSpinner {...props} />)

      const container = screen.getByTestId('loading-spinner')
      expect(container).toBeInTheDocument()
      expect(container).toHaveClass('flex', 'flex-col', 'items-center', 'justify-center')

      // Should default to medium size
      const spinner = within(container).getByRole('generic')
      expect(spinner).toHaveClass('w-8', 'h-8')
    })

    it('should handle null message gracefully', () => {
      render(<LoadingSpinner message={null as any} />)

      const container = screen.getByTestId('loading-spinner')
      expect(container).toBeInTheDocument()

      expect(screen.queryByRole('paragraph')).not.toBeInTheDocument()
    })
  })
})
