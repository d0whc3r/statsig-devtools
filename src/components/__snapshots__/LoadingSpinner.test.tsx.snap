// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LoadingSpinner > should match snapshot with default props 1`] = `
<div>
  <div
    class="flex flex-col items-center justify-center "
    data-testid="loading-spinner"
  >
    <div
      class="animate-spin rounded-full border-2 border-slate-200 w-8 h-8"
      style="border-top-color: #1e40af; border-right-color: #3b82f6;"
    />
  </div>
</div>
`;
